lib/fluent/plugin/in_tail/position_file.rb

```diff
@@ -53,10 +53,12 @@ def [](target_info)
         }
       end
 
-      def unwatch(target_info)
-        if (entry = @map.delete(@follow_inodes ? target_info.ino : target_info.path))
-          entry.update_pos(UNWATCHED_POSITION)
-        end
+      def unwatch_removed_targets(existing_targets)
+        @map.reject { |key, entry|
+          existing_targets.key?(key)
+        }.each_key { |key|
+          unwatch(key)
+        }
```

Comment:
Note: I think we don't need entire exclusive lock here.

Each `unwatch` process has the lock and `try_compact` safely updates the seek positions in one lock.

https://github.com/fluent/fluentd/blob/e12069351a431b30bd84996d60ff9c4a7e78adce/lib/fluent/plugin/in_tail/position_file.rb#L101-L116